homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,112 @@
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    entity_configs = conf.get(CONF_ENTITIES, {})
+
+    def devices():
+        """Devices to be emulated."""
+        yield from get_plug_devices(hass, entity_configs)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    async def start_emulated_kasa(event):
+        for entity_id, entity_config in entity_configs.items():
+            state = hass.states.get(entity_id)
+            if state is None:
+                _LOGGER.warning("Entity not found: %s", entity_id)
+                continue
+        try:
+            await server.start()
+        except OSError as error:
+            _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+        else:
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, start_emulated_kasa)
+
+    return True
+
+
+def get_plug_devices(hass, entity_configs):
+    """Produce list of plug devices from config entities."""
+    for entity_id, entity_config in entity_configs.items():
+        state = hass.states.get(entity_id)
+        if state is None:
+            continue
+        name = entity_config.get(CONF_NAME, state.name)
+
+        if state.state == STATE_ON:
+            if CONF_POWER in entity_config:
+                power_val = entity_config[CONF_POWER]
+                if isinstance(power_val, (float, int)):
+                    power = float(power_val)
+                elif isinstance(power_val, str):
+                    if is_template_string(power_val):
+                        power_val = Template(power_val, hass)
+                        entity_config[CONF_POWER] = power_val
+                        power = float(power_val.async_render())
+                    else:
+                        power = float(hass.states.get(power_val).state)
+                elif isinstance(power_val, Template):
+                    power_val.hass = hass
+                    power = float(power_val.async_render())
+            elif ATTR_CURRENT_POWER_W in state.attributes:
+                power = float(state.attributes[ATTR_CURRENT_POWER_W])
+            else:
+                _LOGGER.warning("No power value defined for: %s", name)
+        else:
+            power = 0.0
```

Comment:
```suggestion
        if CONF_POWER in entity_config:
            power_val = entity_config[CONF_POWER]
            if state.state == STATE_ON and isinstance(power_val, (float, int)):
                power = float(power_val)
            elif isinstance(power_val, str):
                if is_template_string(power_val):
                    power_val = Template(power_val, hass)
                    entity_config[CONF_POWER] = power_val
                    power = float(power_val.async_render())
                else:
                    power = float(hass.states.get(power_val).state)
            elif isinstance(power_val, Template):
                power_val.hass = hass
                power = float(power_val.async_render())
        elif ATTR_CURRENT_POWER_W in state.attributes:
            power = float(state.attributes[ATTR_CURRENT_POWER_W])
        else:
            power = 0.0
            _LOGGER.warning("No power value defined for: %s", name)

        last_changed = state.last_changed.timestamp()
```

Reply:
I think we only need to check STATE_ON if it is a static value.  If its `ATTR_CURRENT_POWER_W` we know its on, and if its a template, we can leave it up to the template.

Reply:
Probably need to cleanup `_LOGGER.warning("No power value defined for: %s", name)` to only warn once as well

Reply:
Working great with the STATE_ON adjustment.  Nice work 🚀 

![Screenshot 2020-09-03 at 9 10 59 PM](https://user-images.githubusercontent.com/663432/92191709-0cdbf400-ee2a-11ea-8c31-81032f2fa9c3.png)

Reply:
When the state is off, it needs to report 0 power usage for a device otherwise static values and sensors that only update when the device is on won't work

Reply:
Also you shouldn't use a value that reports `kwh` as that's going to be power over time and it wants the power rate.  That might be why the state change was needed?

Reply:
> When the state is off, it needs to report 0 power usage for a device otherwise static values and sensors that only update when the device is on won't work

I adjusted my configuration for that, but I don't actually have a switch for my UPS so maybe an `always_on: true` field unless you can think of something better?

```
emulated_kasa:
 entities:
   sensor.ups_kw:
     name: UPS Power
     power: "{{ float(states('sensor.ups_kw')) * 1000 }}"
   switch.pool_pump:
     name: Pool Pump Power
     power: "{{ float(states('sensor.pool_pump_kwh')) * 1000 }}"
```

Reply:
That's a good idea, I didn't consider a sensor without an associated device - I guess if the entity is a sensor, it should just use the sensor value or the power, not try to figure it out from the state